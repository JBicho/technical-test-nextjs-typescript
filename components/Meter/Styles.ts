import styled from 'styled-components';

const MeterContainer = styled.div`
  width: 100%;
  display: flex;
  flex-direction: column;
`;

const StyledMeter = styled.meter`
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  width: 100%;
  border-radius: 10px;
  overflow: hidden;

  &::-webkit-meter-bar {
    background-color: var(--global-accent-color);
    border-radius: 10px;
    background-image: linear-gradient(
      to right,
      var(--global-accent-color),
      var(--global-accent-color)
    );
  }

  &::-webkit-meter-value {
    background-color: #4caf50;
    border-radius: 10px;
  }

  &::-moz-meter-bar {
    background-color: var(--global-accent-color);
    border-radius: 10px;
    background-image: linear-gradient(
      to right,
      var(--global-accent-color),
      var(--global-accent-color)
    );
  }
`;

export { MeterContainer, StyledMeter };
