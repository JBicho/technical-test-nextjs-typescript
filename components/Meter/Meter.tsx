import { MeterContainer, StyledMeter } from './Styles';

export interface MeterProps {
  id: string;
  value: number;
  label: string;
  max: number;
  min?: number;
}

export const Meter = ({ id, value, label, max, min = 0 }: MeterProps) => {
  return (
    <MeterContainer>
      <label htmlFor={id}>{label}</label>
      <StyledMeter id={id} value={value} max={max} min={min} />
    </MeterContainer>
  );
};
