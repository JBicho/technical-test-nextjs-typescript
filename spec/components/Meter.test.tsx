import { render, screen } from '@testing-library/react';
import { Meter, MeterProps } from '../../components/Meter/Meter';
import '../setupTests';

describe('Test Meter Component', () => {
  beforeEach(() => {
    jest.resetModules();
  });

  const meterProps: MeterProps = {
    id: 'someID',
    label: 'Some Label',
    max: 100,
    min: 0,
    value: 50,
  };

  const setupComponent = () => render(<Meter {...meterProps} />);

  it('Renders correctly and matches the snapshot', () => {
    const { asFragment } = setupComponent();

    expect(asFragment()).toMatchSnapshot();
  });

  it('Renders with correct label', () => {
    setupComponent();

    expect(screen.getByText(meterProps.label)).toBeInTheDocument();
  });

  it('Renders meter with correct min, max, and value attributes', () => {
    setupComponent();

    const meterElement = screen.getByRole('meter');

    expect(meterElement).toHaveAttribute('min', '0');
    expect(meterElement).toHaveAttribute('max', '100');
    expect(meterElement).toHaveAttribute('value', '50');
  });
});
