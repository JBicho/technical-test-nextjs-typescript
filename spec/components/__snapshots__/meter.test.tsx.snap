// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Meter Component Renders correctly and matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="Styles__MeterContainer-sc-6555pb-0 kwYyGg"
  >
    <label
      for="someID"
    >
      Some Label
    </label>
    <meter
      class="Styles__StyledMeter-sc-6555pb-1 eeEkfZ"
      id="someID"
      max="100"
      min="0"
      value="50"
    />
  </div>
</DocumentFragment>
`;
